### ISSUE TITLE
Rust build fails on Solaris after: Implement file_lock feature

### ISSUE THREAD

**Issue by psumbera** [rust-lang/rust#132921]
Solaris OS doesn't support `flock()` and thus Rust fails to build after #130999 :
```
d2cdc76256313e8f23f585107c43badb49d5473d is the first bad commit
commit d2cdc76256313e8f23f585107c43badb49d5473d
Author: Christopher Berner [<me@cberner.com>](mailto:me@cberner.com)
Date:   Sun Sep 22 16:58:44 2024 -0700

    Implement file_lock feature
    
    This adds lock(), lock_shared(), try_lock(), try_lock_shared(), and
    unlock() to File gated behind the file_lock feature flag
```

with following error:
```
error: linking with `gcc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0-sysroot/lib/rustlib/x86_64-pc-solaris/bin:/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/bin:/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/bin:/usr/gnu/bin/:/usr/gcc/13/bin:/usr/bin" VSLANG="1033" "gcc" "-m64" "/tmp/rustczm3gUt/symbols.o" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0-rustc/x86_64-pc-solaris/release/deps/rustc_main-f7e62c1025f5dce2.rustc_main.39965e3472e7b77a-cgu.0.rcgu.o" "-Wl,-z,ignore" "-Wl,-Bdynamic" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0-rustc/x86_64-pc-solaris/release/deps/librustc_driver-60ae461c5c39901a.so" "-Wl,-Bstatic" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0-sysroot/lib/rustlib/x86_64-pc-solaris/lib/libcompiler_builtins-579d025a05650947.rlib" "-Wl,-Bdynamic" "-lrt" "-ldl" "-lm" "-lkstat" "-lsocket" "-lz" "-lstdc++" "-lsendfile" "-llgrp" "-lsocket" "-lposix4" "-lpthread" "-lresolv" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lsendfile" "-llgrp" "-L" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0-rustc/x86_64-pc-solaris/release/build/psm-daa576ffce68f706/out" "-L" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0-rustc/x86_64-pc-solaris/release/build/blake3-2e711e7084babad6/out" "-L" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0-rustc/x86_64-pc-solaris/release/build/blake3-2e711e7084babad6/out" "-L" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0-rustc/x86_64-pc-solaris/release/build/rustc_llvm-2c65a157dbd38377/out" "-L" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/llvm/lib" "-o" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0-rustc/x86_64-pc-solaris/release/deps/rustc_main-f7e62c1025f5dce2" "-nodefaultlibs" "-Wl,-z,origin" "-Wl,-rpath,$ORIGIN/../lib"
  = note: Undefined            first referenced
           symbol                  in file
          flock                               /builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0-rustc/x86_64-pc-solaris/release/deps/librustc_driver-60ae461c5c39901a.so
          ld: fatal: symbol referencing errors
          collect2: error: ld returned 1 exit status
```

**Comment by cberner** [comment]
Oh, sorry about that breakage. Is there a way for me to test all the Tier 2 targets to make sure they compile? I can send a PR to fix this, but don't have a Solaris machine for testing

**Comment by psumbera** [comment]
For Solaris `fcntl` locking can be considered. But this will not work within one process. The semantic is different against `flock().`

You should be able to play with Solaris via https://portal.cfarm.net/machines/list/ (cfarm215 and cfarm216). But previous registration is needed.

**Comment by cberner** [comment]
Thanks, I requested an account there.

**Comment by cberner** [comment]
Still waiting for access, so I haven't been able to test [this fix](https://github.com/rust-lang/rust/pull/132977). It seems pretty trivial though, so hopefully it's correct

**Comment by ivmarkov** [comment]
@cberner The new `flock` API also broke ESP-IDF (another Tier 2 target - a single-process embedded one).

As per [here](https://github.com/espressif/esp-idf/issues/2565#issuecomment-430069479), the `flock` API specifically would never be supported on the ESP-IDF as it is a single-process OS and `flock` is all about multi-process locking.

The fix in #132977 should take care of the issue for ESP IDF as well I think, simply because it only enables the `flock` APIs for a few explicitly enumerated platforms, and ESP-IDF is not one of these.

**Comment by cberner** [comment]
Yep, I didn't realize how many OSs use the `unix` implementation directory in the standard library, in my original PR. I think #132977 will fix all of them, like you said
